### PR TITLE
Fixes #18070 - add_record() checks for conflicts with leases now.

### PR DESCRIPTION
**Pull request by dmitri-d**

### CODE REVIEW

**Comment by dmitri-d** [comment]
@timogoebel: updated add_record to check for conflicts with leases.

**Comment by dmitri-d** [comment]
[test]

**Comment by dmitri-d** [comment]
[test]

File: modules/dhcp_common/server.rb [hunk 1]
```diff
@@ -118,44 +118,42 @@ def inspect
...
+      if similar_records.any? {|record| record == to_return}
+        # we already got this record, no need to do anything
+        logger.debug "We already got the same DHCP record - skipping"
+        raise Proxy::DHCP::AlreadyExists
+      end
+
+      unless similar_records.empty?
+        logger.warn "Request to create a conflicting DHCP record"
+        logger.debug "request: #{to_return.inspect}"
+        logger.debug "existing:   #{similar_records.inspect}"
```
Comment (timogoebel) [inline modules/dhcp_common/server.rb]: Too many whitespaces between `:` and `#`

File: modules/dhcp_common/record.rb [hunk 2]
```diff
@@ -34,25 +33,19 @@ def [] opt
...
 
-    # compare between incoming request and our existing record
-    # if our record has all requested attributes then we say they are comparable
-    def self.compare_options(record, request)
-      request.each do |k,v|
-        return false if record[k.to_sym] != v
-      end
-      true
+    def ==(other)
+      ip == other.ip && mac == other.mac && subnet == other.subnet && deleteable? == other.deleteable? && options == other.options
```
Comment (timogoebel) [inline modules/dhcp_common/record.rb]: Should we check if other is really a record before comparison? Same for `Proxy::DHCP::Lease`.

File: modules/dhcp_common/server.rb [hunk 3]
```diff
@@ -118,44 +118,42 @@ def inspect
     def add_record options = {}
       related_macs = options.delete("related_macs") || []
       logger.debug "Ignoring duplicates for macs: #{related_macs.inspect}" unless related_macs.empty?
-      options = clean_up_add_record_parameters(options)
+      name, ip_address, mac_address, subnet_address, options = clean_up_add_record_parameters(options)
 
-      validate_ip(options[:ip])
-      validate_mac(options[:mac])
-      raise(Proxy::DHCP::Error, "Must provide hostname") unless options[:hostname]
```
Comment (timogoebel) [inline modules/dhcp_common/server.rb:125]: Is this safe to remove because we use "name" as a fallback? Should we add a check if name is set?

**Review by timogoebel — COMMENTED** [review]
@witlessbird : Thanks. I've left some inline comments!

**Comment by dmitri-d** [comment]
- restored name presence check in add_record method
- fixed formatting in https://github.com/theforeman/smart-proxy/pull/498#discussion-diff-97348267R143
- rewrote equality methods to remove duplication

**Review by timogoebel — APPROVED** [review]
Looks good and works great, thanks @witlessbird.

**Pull request merged**